### DISCUSSION TITLE
typescript hangs when compiling plugin with tsc 4.5

### DISCUSSION THREAD

**Discussion by stefanprobst** [unifiedjs/unified#175] [Q&A] [answered]
i'm not sure if this s a typescript or a unified issue, but i'm having trouble compiling a simple plugin with ts 4.5 (tsc just hangs):

could somebody try building this example: https://github.com/stefanprobst/issue-tsc

notes:
- when removing the third arg (`parent`) to the visitor function, this compile sfine
- ts 4.4.4 also seems to work

**Comment by ChristianMurphy** [comment, marked as answer]
Given that it works in 4.4 and breaks in 4.5 I tend do think this is a TS bug. 🤔 
Maybe an edge case with https://devblogs.microsoft.com/typescript/announcing-typescript-4-5/#tailrec-conditional ?

But we may end up having to update unified if the answer is similar to https://github.com/microsoft/TypeScript/issues/26314 or https://github.com/microsoft/TypeScript/issues/25942#issuecomment-413976729

**Reply by stefanprobst** [reply]
yes, filed a bug report, and it looks like it is a ts issue.

**Reply by ChristianMurphy** [reply]
Thanks, linking for reference https://github.com/microsoft/TypeScript/issues/46900

**Reply by styxlab** [reply]
Just experiencing the same issue. I changed typescript to `4.4.4` in package.json but that did not help. `tsc -v` shows `4.5.2` do I need to downgrade that too?

**Reply by styxlab** [reply]
Okay, downgrading tsc to `4.4.4` helped. Thanks for this bug report which helped me save a lot of time!

**Reply by SilvaQ** [reply]
Pure JS project into the library is full of CPU, how to deal with?

**Reply by ChristianMurphy** [reply]
This has been resolved in TypeScript 4.6 https://devblogs.microsoft.com/typescript/announcing-typescript-4-6/#improved-recursion-depth-checks
The fix has also been backported to 4.5

> You may already have this change because it was cherry-picked into TypeScript 4.5.3

**Comment by ChristianMurphy** [comment]
This has been resolved in TypeScript 4.6 https://devblogs.microsoft.com/typescript/announcing-typescript-4-6/#improved-recursion-depth-checks
The fix has also been backported to 4.5

> You may already have this change because it was cherry-picked into TypeScript 4.5.3
